packages/web-components/fast-element/src/components/element-controller.ts

```diff
@@ -470,3 +468,117 @@ export class ElementController<TElement extends HTMLElement = HTMLElement>
         ));
     }
 }
+
+/**
+ * Converts a styleTarget into the operative target. When the provided target is an Element
+ * that is a FASTElement, the function will return the ShadowRoot for that element. Otherwise,
+ * it will return the root node for the element.
+ * @param target
+ * @returns
+ */
+function normalizeStyleTarget(target: StyleTarget): StyleTarget {
+    if ("adoptedStyleSheets" in target) {
+        return target;
+    } else {
+        return (
+            (getShadowRoot(target as any) as null | StyleTarget) ??
+            (target.getRootNode() as any)
+        );
+    }
+}
+
+// Default StyleStrategy implementations are defined in this module because they
+// require access to element shadowRoots, and we don't want to leak shadowRoot
+// objects out of this module.
+/**
+ * https://wicg.github.io/construct-stylesheets/
+ * https://developers.google.com/web/updates/2019/02/constructable-stylesheets
+ *
+ * @internal
+ */
+export class AdoptedStyleSheetsStrategy implements StyleStrategy {
+    private static styleSheetCache = new Map<string, CSSStyleSheet>();
+    /** @internal */
+    public readonly sheets: CSSStyleSheet[];
+
+    public constructor(styles: (string | CSSStyleSheet)[]) {
+        const styleSheetCache = AdoptedStyleSheetsStrategy.styleSheetCache;
+        this.sheets = styles.map((x: string | CSSStyleSheet) => {
+            if (x instanceof CSSStyleSheet) {
+                return x;
+            }
+
+            let sheet = styleSheetCache.get(x);
+
+            if (sheet === void 0) {
+                sheet = new CSSStyleSheet();
+                (sheet as any).replaceSync(x);
+                styleSheetCache.set(x, sheet);
+            }
+
+            return sheet;
+        });
+    }
+
+    public addStylesTo(target: StyleTarget): void {
+        const t = normalizeStyleTarget(target);
+        t.adoptedStyleSheets = [...t.adoptedStyleSheets!, ...this.sheets];
+    }
+
+    public removeStylesFrom(target: StyleTarget): void {
+        const t = normalizeStyleTarget(target);
+        const sheets = this.sheets;
+        t.adoptedStyleSheets = t.adoptedStyleSheets!.filter(
+            (x: CSSStyleSheet) => sheets.indexOf(x) === -1
+        );
+    }
+}
+
+let id = 0;
+const nextStyleId = (): string => `fast-${++id}`;
+function usableStyleTarget(target: StyleTarget): StyleTarget {
+    return target === document ? document.body : target;
+}
+/**
+ * @internal
+ */
+export class StyleElementStrategy implements StyleStrategy {
+    private readonly styleClass: string;
+
+    public constructor(private readonly styles: string[]) {
+        this.styleClass = nextStyleId();
+    }
+
+    public addStylesTo(target: StyleTarget): void {
+        target = usableStyleTarget(normalizeStyleTarget(target));
+
+        const styles = this.styles;
+        const styleClass = this.styleClass;
+
+        for (let i = 0; i < styles.length; i++) {
```

Comment:
```suggestion
        for (let i = 0, stylesLength = styles.length; i < stylesLength; i++) {
```

Reply:
Going to leave this as-is mostly so I don't have to rebuild.
I just moved this code from another file, so it was pre-existing.